common/display/displayqueue.cpp

```diff
@@ -340,6 +356,50 @@ void DisplayQueue::HandleExit() {
   compositor_.Reset();
 }
 
+bool DisplayQueue::SetDrmPipeScalePropery(drmModeAtomicReqPtr pset,
+                                          uint32_t src_width,
+                                          uint32_t src_height, uint32_t dst_x,
+                                          uint32_t dst_y, uint32_t dst_w,
+                                          uint32_t dst_h) {
+  if (pipe_scale_mode_ == 0)
+    return false;  // pipe scaler is not supported
+ 
+  int success = drmModeAtomicAddProperty(pset, crtc_id_, pipe_scale_mode_,
+                                        DRM_MODE_SCALE_CENTER) < 0;
+  
+  success |=
+      drmModeAtomicAddProperty(pset, crtc_id_, pipe_src_w_, src_width) < 0;
+  success |=
+      drmModeAtomicAddProperty(pset, crtc_id_, pipe_src_h_, src_height) < 0;
+  success |= drmModeAtomicAddProperty(pset, crtc_id_, pipe_dst_x_, dst_x) < 0;
+  success |= drmModeAtomicAddProperty(pset, crtc_id_, pipe_dst_y_, dst_y) < 0;
+  success |= drmModeAtomicAddProperty(pset, crtc_id_, pipe_dst_w_, dst_w) < 0;
+  success |= drmModeAtomicAddProperty(pset, crtc_id_, pipe_dst_h_, dst_h) < 0;
+
+  if (success) {
+    ETRACE("Could not update properties for crtc scaler for crtc_id: %d",
+           crtc_id_);
+    return false;
+  }
+  
+  return true;
+}
+
+void DisplayQueue::ApplyPipeScaler(
+    DisplayPlaneStateList& current_composition_planes,
+    drmModeAtomicReqPtr pset) {
+  // here we define a simple strategy about how to use piple scaler
+  // single plane and single layer
```

Comment:
We need to have this check in ValidateLayers in DisplayPlaneManager, ensure in this case test commit passes so we dont fallback to 3D composition. Otherwise this here wont make sense. Perhaps u can make a help class for pipe scaling which will be owned by DisplayQueue and pointer passed to DisplayPlaneManager in ValidateLayers call.